gpu4pyscf/nac/mecp.py

```diff
@@ -0,0 +1,243 @@
+        grad_method = self.td.nuc_grad_method()
+        g1 = grad_method.kernel(state=self.states[0])
+        g2 = grad_method.kernel(state=self.states[1])
+
+        # 2. Define the branching space vectors x1 and x2
+        # x1 is the gradient difference vector
+        x1 = g1 - g2
+        x1_norm_val = np.linalg.norm(x1)
+        x1_norm_vec = x1 / x1_norm_val if x1_norm_val > 1e-9 else np.zeros_like(x1)
+
+        # For minimum energy crossing point, x2 is the non-adiabatic coupling vector
+        if self.crossing_type == 'n-2':
+            nac_method = self.td.nac_method()
+            nac_method.states=self.states
+            nac_vect = nac_method.kernel()
+            x2 = nac_vect[1]
+            x2_norm_val = np.linalg.norm(x2)
+            x2_norm_vec = x2 / x2_norm_val if x2_norm_val > 1e-9 else np.zeros_like(x2)
+            
+            # Project the gradient of second state onto the seam space and take the component outside
+            natom = g2.shape[0]
+            g_on_plane = project_on_plane_lstsq(g2, x1_norm_vec, x2_norm_vec)
+            g_on_plane = g_on_plane.reshape(natom, 3)
+            g_proj = g2 - g_on_plane
+        
+        elif self.crossing_type == 'n-1':
+            raise NotImplementedError("n-1 crossing type not implemented yet.")
+
+        # 3. Calculate component of the effective gradient driving states to degeneracy
+        f = (E1 - E2) * x1_norm_vec  # Note: The original paper has (E1-E2)
+        
+        # 4. Total effective gradient
+        g_bar = g_proj + f
+
+        self.log.info(f"  ||Seam Grad (g_proj)||: {np.linalg.norm(g_proj):.6f}")
+        self.log.info(f"  ||Degeneracy Grad (f)||: {np.linalg.norm(f):.6f}")
+        self.log.info(f"  ||Total Effective Grad||: {np.linalg.norm(g_bar):.6f}")
+        self.log.info("----------------------------------------------------------------")
+        
+        # The optimizer minimizes a single energy value. We provide the average.
+        energy_for_optimizer = (E1 + E2) / 2.0
+        
+        # Cache the results for the current geometry
+        self._last_geom = np.copy(current_geom)
+        self._last_energy = energy_for_optimizer
+        self._last_grad = g_bar
+        
+        return energy_for_optimizer, g_bar
+
+    def optimize(self, geom=None, **kwargs):
+        """
+        Runs the geometry optimization using pyscf.geomopt.geometric_solver.
+        
+        Args:
+            geom (str or np.ndarray): Initial geometry. If None, uses the geometry
+                                      from the molecule in the MCSCF object.
+            **kwargs: Additional keyword arguments to pass to the optimizer.
+                      e.g., max_cycle=100, prefix='traj.xyz'
+            TODO: kwargs seem not work!
+        """
+        if geom is not None:
+            self.mol.atom = geom
+
+        # Create the wrapper object that geometric_solver will use
+        solver_wrapper = self._SolverWrapper(self)
```

Comment:
the solver_wrapper can be created as a pyscf.lib.GradScanner instance. The geometric_solver function will execute this as `solver_wrapper(new_goemetry)` . The `td_scanner` mentioned in the `get_eff_energy_and_gradient` function can be called in the scanner wrapper.

Adaptation to ris can be executed in `solver_wrapper()`. The mean-field object in ris can be created as `mf.as_scanner()`. and mf calculation should be executed through the mf_scanner. A `reset` method for ris should be added, to cleanup intermediates for the old geometry. and the ris' reset() should also be called within the `solver_wrapper()`